src/main/java/bio/terra/service/filedata/azure/AzureSynapsePdao.java

```diff
@@ -485,12 +524,228 @@ public void createSnapshotRowIdsParquetFile(
+            .stream()
+            .map(Column::toSynapseColumn)
+            .collect(Collectors.toList());
+    ST sqlCreateSnapshotTableTemplate = new ST(createSnapshotTableByAssetTemplate);
+    ST queryTemplate =
+        generateSnapshotParquetCreateQuery(
+            sqlCreateSnapshotTableTemplate,
+            IngestUtils.getSourceDatasetParquetFilePath(rootTable.getTable().getName()),
+            rootTable.getTable().getName(),
+            snapshotId,
+            IngestUtils.getSnapshotSliceParquetFilePath(snapshotId, rootTableName, "root"),
+            datasetDataSourceName,
+            snapshotDataSourceName,
+            columns);
+
+    AssetColumn rootColumn = assetSpec.getRootColumn();
+    queryTemplate.add("rootColumn", rootColumn.getDatasetColumn().getName());
+    String query = queryTemplate.render();
+    MapSqlParameterSource params =
+        new MapSqlParameterSource().addValue("rootValues", requestModel.getRootValues());
+    int rows;
+    try {
+      rows = synapseJdbcTemplate.update(query, params);
+    } catch (DataAccessException ex) {
+      logger.error("Error while creating root table parquet file on snapshot by asset", ex);
+      throw new PdaoException(
+          "Unable to create parquet file for the root table. This is most likely because the source dataset table is empty.  See exception details if this does not appear to be the case. Exception cause: "
+              + ex.getCause().getMessage());
+    }
+
+    tableRowCounts.put(rootTableName, (long) rows);
+    if (rows == 0) {
+      throw new PdaoException(
+          "Snapshot by Asset - No rows included from root table due to root values that were specified. Cannot create entirely empty snapshot.");
+    } else {
+      logger.info("Snapshot by Asset - {} rows included in root table {}", rows, rootTableName);
+    }
+
+    // Then walk relationships
+    UUID rootTableId = rootTable.getTable().getId();
+    List<WalkRelationship> walkRelationships = WalkRelationship.ofAssetSpecification(assetSpec);
+    walkRelationships(
+        snapshotId,
+        assetSpec,
+        datasetDataSourceName,
+        snapshotDataSourceName,
+        rootTableId,
+        walkRelationships,
+        tableRowCounts);
+
+    return tableRowCounts;
+  }
+
+  public void walkRelationships(
+      UUID snapshotId,
+      AssetSpecification assetSpec,
+      String datasetDataSourceName,
+      String snapshotDataSourceName,
+      UUID startTableId,
+      List<WalkRelationship> walkRelationships,
+      Map<String, Long> tableRowCounts)
+      throws SQLException {
+    for (WalkRelationship relationship : walkRelationships) {
+      if (relationship.visitRelationship(startTableId)) {
+        createSnapshotParquetFilesByRelationship(
+            snapshotId,
+            assetSpec,
+            relationship,
+            datasetDataSourceName,
+            snapshotDataSourceName,
+            tableRowCounts);
+        walkRelationships(
```

Comment:
Quick q: do we need to do cycle detection?  (this might not be the place to do it).  I don't know if we do that in GCP?

Reply:
I'm re-using the logic from GCP.  On processing a relationship, we're marking the relationship as visited so that it doesn't get into an infinite loop. It's discussed in [this comment](https://github.com/DataBiosphere/jade-data-repo/blob/develop/src/main/java/bio/terra/service/tabulardata/google/bigquery/BigQuerySnapshotPdao.java#L919).